lib/aspects/user-only-permission-boundary.ts

```diff
@@ -0,0 +1,17 @@
+import * as cdk from "aws-cdk-lib";
+import * as iam from "aws-cdk-lib/aws-iam";
+import { IConstruct } from "constructs";
+
+export class UserPermissionBoundary implements cdk.IAspect {
+  private readonly policy: iam.ManagedPolicy;
+  constructor(boundary: iam.ManagedPolicy) {
+    this.policy = boundary;
+  }
+
+  visit(node: IConstruct): void {
+    if (!(node instanceof iam.User)) {
+      return;
+    }
+    iam.PermissionsBoundary.of(node).apply(this.policy);
+  }
```

Comment:
When is this `visit` method actually being called, during `cdk deploy`? 

Also, I believe you mentioned it already, but want to make sure I understand, but here the users IAM gets an additional policy to invoke the `api-gw`, along with any other permissions (if any) that are present for that particular user?

Reply:
The CDK itself leverages the [Visitor pattern](https://refactoring.guru/design-patterns/visitor) through [Aspects](https://docs.aws.amazon.com/cdk/v2/guide/aspects.html). So we don't call `visit` but it gets applied by the Aspects framework.

Reply:
I think Julius was asking whether it the aspect was actually applied during `synth` or `deploy`.

Reply:
Thanks for pointing out documentation that helps clear things up.